async_ssh2_russh/src/promise.rs

```diff
@@ -0,0 +1,195 @@
+        self.resolve(value).unwrap_or_else(|_| panic!("already resolved"));
+    }
+
+    /// Resolves the promise with the given value, returning an error if the promise has already been resolved.
+    pub fn resolve(&mut self, value: T) -> Result<(), T> {
+        let Some(inner) = self.inner.take() else {
+            return Err(value);
+        };
+
+        if let Some(inner) = inner.upgrade() {
+            // SAFETY: `&mut self: Resolve` has exclusive access to `resolve` once.
+            unsafe {
+                inner.resolve(Some(value));
+            }
+        }
+        Ok(())
+    }
+}
+impl<T> Drop for Resolve<T> {
+    fn drop(&mut self) {
+        if let Some(inner) = self.inner.take().and_then(|weak| weak.upgrade()) {
+            // SAFETY: `&mut self: Resolve` has exclusive access to call resolve once. Because we use `inner.take()`, we
+            // know `Self::resolve` was not called.
+            unsafe {
+                inner.resolve(None);
+            }
+        }
+    }
+}
+
+/// Flag for when the [`Resolve`] will no longer change the value of the [`Inner`], because it was either resolved or
+/// dropped.
+const FLAG_COMPLETED: u8 = 0b01;
+/// Flag for when the value is set and can be read.
+const FLAG_VALUE_SET: u8 = 0b10;
+
+/// Any thread with an `&self` may access the `value` field according the following rules:
+///
+///  1. Iff `flag & FLAG_COMPLETED` is false, the `value` field may be modified by the owning `Resolve` once.
+///  2. Iff `flag & FLAG_VALUE_SET` is true, the `value` field may be accessed immutably by any thread.
+///
+/// If `flag & FLAG_COMPLETED` is true, but `flag & FLAG_VALUE_SET` is false, then the owning `Resolve` was dropped
+/// before the value was set.
+struct Inner<T> {
+    flag: AtomicU8,
+    value: UnsafeCell<MaybeUninit<T>>,
+    wakers: Mutex<Vec<Waker>>,
+}
+impl<T> Inner<T> {
+    const fn new() -> Self {
+        Self {
+            flag: AtomicU8::new(0),
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            wakers: Mutex::new(Vec::new()),
+        }
+    }
+
+    fn poll_get<'a>(&'a self, cx: &mut Context<'_>) -> Poll<Option<&'a T>> {
+        if let Some(value_opt) = self.get() {
+            return Poll::Ready(value_opt);
+        }
+        {
+            // Acquire lock.
+            let mut wakers = self.wakers.lock().unwrap();
+            // Check again in case of race condition with resolver.
+            if let Some(value_opt) = self.get() {
+                return Poll::Ready(value_opt);
+            }
+            // Add the current waker to the list of wakers.
+            wakers.push(cx.waker().clone());
+        }
+        Poll::Pending
+    }
+
+    fn get(&self) -> Option<Option<&T>> {
+        // Using acquire ordering so any threads that read a true from this
+        // atomic is able to read the value.
+        let flag = self.flag.load(Ordering::Acquire);
+        let resolved = 0 != flag & FLAG_COMPLETED;
```

Comment:
a bit compact to read. at minimum I encourage a bit of parenthesization for readability even if it's correct without.

also maybe name the variable `completed` for clarity?